core/src/main/java/org/elasticsearch/common/xcontent/XContentParserUtils.java

```diff
@@ -122,22 +123,29 @@ public static Object parseStoredFieldsValue(XContentParser parser) throws IOExce
      * @param delimiter   the delimiter to use to splits the field's name
      * @param objectClass the object class of the object to parse
      * @param <T>         the type of the object to parse
+     * @param lenient     set to <tt>true</tt> if the method should throw a ParsingException in no delimeter was found
      * @return the parsed object
      * @throws IOException if anything went wrong during parsing or if the type or name cannot be derived
      *                     from the field's name
      */
-    public static <T> T parseTypedKeysObject(XContentParser parser, String delimiter, Class<T> objectClass) throws IOException {
+    public static <T> Optional<T> parseTypedKeysObject(XContentParser parser, String delimiter, Class<T> objectClass, boolean lenient)
```

Comment:
I agree, passing in collections that are added to is not beautiful, but I still prefer it over having Optional or null checks in all the callers. Maybe like Tanguy suggested we would provide a Consumer rather than the list.